### ISSUE TITLE
GPU mem leak when export with Dict output [JIT]

### ISSUE THREAD

**Issue by Godricly** [pytorch/pytorch#33437]
## 🐛 Bug
The Dict output export has mem leak if called repeatly. The list one is stable.

## To Reproduce

1. run the script.
1. OOM is expected after 100 iters for dict.

```
import torch
from torch import nn


class Demo1(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(3, 20, 3)
        self.conv2 = nn.Conv2d(3, 20, 3)

    def forward(self, x):
        return [self.conv1(x), self.conv2(x)]


class Demo2(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(3, 20, 3)
        self.conv2 = nn.Conv2d(3, 20, 3)

    def forward(self, x):
        return {'out1': self.conv1(x), 'out2': self.conv2(x)}


def test_demo(model):
    inputs = torch.zeros(2, 3, 512, 512)
    if torch.cuda.is_available():
        model = model.cuda()
        inputs = inputs.cuda()

    for i in range(100):
        model.train()
        torch.onnx.export(model, inputs,
                          'debug.onnx',
                          input_names=['data'],
                          opset_version=11,
                          verbose=True,
                          dynamic_axes={'data': {0: 'batch', 2: 'width', 3: 'height'}})
        torch.cuda.empty_cache()
        print(torch.cuda.memory_summary(device=None, abbreviated=False))

if __name__ == '__main__':
    model1 = Demo1()
    model2 = Demo2()
    print('demo1')
    test_demo(model1)
    print('demo2')
    test_demo(model2)
```
## Expected behavior
increasing memory
```
|===========================================================================|
|                  PyTorch CUDA memory summary, device ID 0                 |
|---------------------------------------------------------------------------|
|            CUDA OOMs: 0            |        cudaMalloc retries: 0         |
|===========================================================================|
|        Metric         | Cur Usage  | Peak Usage | Tot Alloc  | Tot Freed  |
|---------------------------------------------------------------------------|
| Allocated memory      |    6806 MB |    6813 MB |   16473 MB |    9667 MB |
|       from large pool |    6806 MB |    6813 MB |   16472 MB |    9666 MB |
|       from small pool |       0 MB |       0 MB |       1 MB |       1 MB |
|---------------------------------------------------------------------------|
| Active memory         |    6806 MB |    6813 MB |   16473 MB |    9667 MB |
|       from large pool |    6806 MB |    6813 MB |   16472 MB |    9666 MB |
|       from small pool |       0 MB |       0 MB |       1 MB |       1 MB |
|---------------------------------------------------------------------------|
| GPU reserved memory   |    6822 MB |    6822 MB |   14822 MB |    8000 MB |
|       from large pool |    6820 MB |    6820 MB |   14820 MB |    8000 MB |
|       from small pool |       2 MB |       2 MB |       2 MB |       0 MB |
|---------------------------------------------------------------------------|
| Non-releasable memory |   16372 KB |   16378 KB |    1691 MB |    1675 MB |
|       from large pool |   14336 KB |   14336 KB |    1688 MB |    1674 MB |
|       from small pool |    2036 KB |    2045 KB |       3 MB |       1 MB |
|---------------------------------------------------------------------------|
| Allocations           |     179    |     185    |    1675    |    1496    |
|       from large pool |     171    |     173    |     927    |     756    |
|       from small pool |       8    |      12    |     748    |     740    |
|---------------------------------------------------------------------------|
| Active allocs         |     179    |     185    |    1675    |    1496    |
|       from large pool |     171    |     173    |     927    |     756    |
|       from small pool |       8    |      12    |     748    |     740    |
|---------------------------------------------------------------------------|
| GPU reserved segments |     172    |     172    |     372    |     200    |
|       from large pool |     171    |     171    |     371    |     200    |
|       from small pool |       1    |       1    |       1    |       0    |
|---------------------------------------------------------------------------|
| Non-releasable allocs |       2    |       2    |       3    |       1    |
|       from large pool |       1    |       1    |       2    |       1    |
|       from small pool |       1    |       1    |       1    |       0    |
|===========================================================================|

|===========================================================================|
|                  PyTorch CUDA memory summary, device ID 0                 |
|---------------------------------------------------------------------------|
|            CUDA OOMs: 0            |        cudaMalloc retries: 0         |
|===========================================================================|
|        Metric         | Cur Usage  | Peak Usage | Tot Alloc  | Tot Freed  |
|---------------------------------------------------------------------------|
| Allocated memory      |    6886 MB |    6893 MB |   16562 MB |    9676 MB |
|       from large pool |    6886 MB |    6893 MB |   16561 MB |    9675 MB |
|       from small pool |       0 MB |       0 MB |       1 MB |       1 MB |
|---------------------------------------------------------------------------|
| Active memory         |    6886 MB |    6893 MB |   16562 MB |    9676 MB |
|       from large pool |    6886 MB |    6893 MB |   16561 MB |    9675 MB |
|       from small pool |       0 MB |       0 MB |       1 MB |       1 MB |
|---------------------------------------------------------------------------|
| GPU reserved memory   |    6902 MB |    6902 MB |   14902 MB |    8000 MB |
|       from large pool |    6900 MB |    6900 MB |   14900 MB |    8000 MB |
|       from small pool |       2 MB |       2 MB |       2 MB |       0 MB |
|---------------------------------------------------------------------------|
| Non-releasable memory |   16372 KB |   16378 KB |    1700 MB |    1684 MB |
|       from large pool |   14336 KB |   14336 KB |    1697 MB |    1683 MB |
|       from small pool |    2036 KB |    2045 KB |       3 MB |       1 MB |
|---------------------------------------------------------------------------|
| Allocations           |     181    |     187    |    1684    |    1503    |
|       from large pool |     173    |     175    |     932    |     759    |
|       from small pool |       8    |      12    |     752    |     744    |
|---------------------------------------------------------------------------|
| Active allocs         |     181    |     187    |    1684    |    1503    |
|       from large pool |     173    |     175    |     932    |     759    |
|       from small pool |       8    |      12    |     752    |     744    |
|---------------------------------------------------------------------------|
| GPU reserved segments |     174    |     174    |     374    |     200    |
|       from large pool |     173    |     173    |     373    |     200    |
|       from small pool |       1    |       1    |       1    |       0    |
|---------------------------------------------------------------------------|
| Non-releasable allocs |       2    |       2    |       3    |       1    |
|       from large pool |       1    |       1    |       2    |       1    |
|       from small pool |       1    |       1    |       1    |       0    |
|===========================================================================|

|===========================================================================|
|                  PyTorch CUDA memory summary, device ID 0                 |
|---------------------------------------------------------------------------|
|            CUDA OOMs: 0            |        cudaMalloc retries: 0         |
|===========================================================================|
|        Metric         | Cur Usage  | Peak Usage | Tot Alloc  | Tot Freed  |
|---------------------------------------------------------------------------|
| Allocated memory      |    6966 MB |    6973 MB |   16651 MB |    9685 MB |
|       from large pool |    6966 MB |    6973 MB |   16650 MB |    9684 MB |
|       from small pool |       0 MB |       0 MB |       1 MB |       1 MB |
|---------------------------------------------------------------------------|
| Active memory         |    6966 MB |    6973 MB |   16651 MB |    9685 MB |
|       from large pool |    6966 MB |    6973 MB |   16650 MB |    9684 MB |
|       from small pool |       0 MB |       0 MB |       1 MB |       1 MB |
|---------------------------------------------------------------------------|
| GPU reserved memory   |    6982 MB |    6982 MB |   14982 MB |    8000 MB |
|       from large pool |    6980 MB |    6980 MB |   14980 MB |    8000 MB |
|       from small pool |       2 MB |       2 MB |       2 MB |       0 MB |
|---------------------------------------------------------------------------|
| Non-releasable memory |   16372 KB |   16378 KB |    1709 MB |    1693 MB |
|       from large pool |   14336 KB |   14336 KB |    1706 MB |    1692 MB |
|       from small pool |    2036 KB |    2045 KB |       3 MB |       1 MB |
|---------------------------------------------------------------------------|
| Allocations           |     183    |     189    |    1693    |    1510    |
|       from large pool |     175    |     177    |     937    |     762    |
|       from small pool |       8    |      12    |     756    |     748    |
|---------------------------------------------------------------------------|
| Active allocs         |     183    |     189    |    1693    |    1510    |
|       from large pool |     175    |     177    |     937    |     762    |
|       from small pool |       8    |      12    |     756    |     748    |
|---------------------------------------------------------------------------|
| GPU reserved segments |     176    |     176    |     376    |     200    |
|       from large pool |     175    |     175    |     375    |     200    |
|       from small pool |       1    |       1    |       1    |       0    |
|---------------------------------------------------------------------------|
| Non-releasable allocs |       2    |       2    |       3    |       1    |
|       from large pool |       1    |       1    |       2    |       1    |
|       from small pool |       1    |       1    |       1    |       0    |
|===========================================================================|

```
## Environment
Collecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.15.3

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: GeForce GTX 1080
Nvidia driver version: 430.50
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.3

Versions of relevant libraries:
[pip3] numpy==1.17.2
[pip3] torch==1.4.0
[pip3] torchvision==0.5.0
[conda] Could not collect
## Additional context

cc @suo @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof

**Comment by eellison** [comment]
Does this only occur when the model is exported (not when it is just run) ?

**Comment by Godricly** [comment]
Whats the difference？ the demo does nothing but export

**Comment by Godricly** [comment]
@eellison Is there any  update for this one?

**Comment by eellison** [comment]
@driazati was looking into it

**Comment by Godricly** [comment]
This demo still bugs in recent 1.6.0 release.

**Comment by Godricly** [comment]
could anyone fix this please?

**Comment by mnidza** [comment]
Why was this closed? I can still repro with version 2.0.1.

**Comment by Godricly** [comment]
I closed this issue to clean up issues I created. If anyone can fix this then it is worthy to reopen.
